### PR TITLE
Split up the two parts of cmdlinet::parse into separate functions

### PR DESCRIPTION
**Pull request by hannes-steffenhagen-diffblue**
`cmdlinet::parse` first parses the command line specification, and then the command line arguments themselves. This PR splits those two bits up into their own functions (this just moves code around, the only reason the functions look slightly different is because of `clang-format`).

- [ ] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: src/util/cmdline.h [hunk 1]
```diff
@@ -112,6 +112,8 @@ class cmdlinet
     {}
   };
 
+  void parse_optstring(const char *optstring);
```
Comment (thomasspriggs) [inline src/util/cmdline.h:174]: :pick: Missing doxygen. Only a nitpick, as it was previously missing from the `parse` member function.

File: src/util/cmdline.h [hunk 2]
```diff
@@ -113,6 +113,7 @@ class cmdlinet
   };
 
   void parse_optstring(const char *optstring);
+  bool parse_arguments(int argc, const char **argv);
```
Comment (thomasspriggs) [inline src/util/cmdline.h:180]: :pick: Missing doxygen. Only a nitpick, as it was previously missing from the `parse` member function.

**Review by thomasspriggs — APPROVED** [review]

**Comment by thomasspriggs** [comment]
Your commits have empty message bodies. They should really have a rationale for the changes are being made.

**Review by NlightNFotis — APPROVED** [review]
Cool, nice maintenance work 👍 

Would also like if we could get docstrings for the functions introduced, but it's a nitpick for me as well - just the cherry on top, really.

**Comment by hannes-steffenhagen-diffblue** [comment]
@thomasspriggs @NlightNFotis I have added a bit of documentation, if you would check if that works for you?

File: src/util/cmdline.h [hunk 3]
```diff
@@ -20,6 +20,58 @@ Author: Daniel Kroening, kroening@kroening.com
...
+  /// Parses a commandline according to a specification given in optstring.
+  /// \param argc How many arguments there are.
+  /// \param argv An array of C strings.
+  ///             The 0th element is assumed to be the name of the command as
+  ///             it was invoked (e.g. /usr/bin/cmake) and is ignored. It is
+  ///             further assumed the array holds argc+1 elements with the C
+  ///             string at index argc being a terminating null pointer.
+  ///             This parameter is parsed based on optstring.
+  /// \param optstring A specification of allowed command line options.
+  ///                  This is a C string container any number of single
```
Comment (thomasspriggs) [inline src/util/cmdline.h:32]: :pick: container -> containing

**Review by thomasspriggs — APPROVED** [review]

File: src/util/cmdline.cpp [hunk 4]
```diff
@@ -249,6 +210,49 @@ cmdlinet::option_namest cmdlinet::option_names() const
 {
   return option_namest{*this};
 }
+void cmdlinet::parse_optstring(const char *optstring)
```
Comment (tautschnig) [inline src/util/cmdline.cpp]: Nit pick: a blank line would be nice to have.

File: src/util/cmdline.h [hunk 5]
```diff
@@ -20,6 +20,58 @@ Author: Daniel Kroening, kroening@kroening.com
 class cmdlinet
 {
 public:
+  /// Parses a commandline according to a specification given in optstring.
```
Comment (tautschnig) [inline src/util/cmdline.h]: I'd suggest `\p optstring` to clarify to an extent what this is referring to.

File: src/util/cmdline.h [hunk 6]
```diff
@@ -20,6 +20,58 @@ Author: Daniel Kroening, kroening@kroening.com
...
+  /// \param argc How many arguments there are.
+  /// \param argv An array of C strings.
+  ///             The 0th element is assumed to be the name of the command as
+  ///             it was invoked (e.g. /usr/bin/cmake) and is ignored. It is
+  ///             further assumed the array holds argc+1 elements with the C
+  ///             string at index argc being a terminating null pointer.
+  ///             This parameter is parsed based on optstring.
```
Comment (tautschnig) [inline src/util/cmdline.h]: Again, `\p optstring`, and s/parameter/argument/ ("argv" is a parameter, but what gets parsed is the actual argument).

File: src/util/cmdline.h [hunk 7]
```diff
@@ -20,6 +20,58 @@ Author: Daniel Kroening, kroening@kroening.com
 class cmdlinet
 {
 public:
+  /// Parses a commandline according to a specification given in optstring.
+  /// \param argc How many arguments there are.
+  /// \param argv An array of C strings.
+  ///             The 0th element is assumed to be the name of the command as
+  ///             it was invoked (e.g. /usr/bin/cmake) and is ignored. It is
+  ///             further assumed the array holds argc+1 elements with the C
```
Comment (tautschnig) [inline src/util/cmdline.h]: `\p argc+1`

File: src/util/cmdline.h [hunk 8]
```diff
@@ -20,6 +20,58 @@ Author: Daniel Kroening, kroening@kroening.com
...
+  ///                  This is a C string container any number of single
+  ///                  characters other than '(', ')' or ':' signifying a
+  ///                  "short" option consisting of just that character, or
+  ///                  names consisting of any characters other than ')'
+  ///                  surrounded by a matching pair of '(' and ')' signifying a
+  ///                  "long" option with the name being the string between '('
+  ///                  and ')', both of which can be optionally followed by a
+  ///                  single ':' indicating that the option takes a parameter,
+  ///                  if not present it does not. Parameters must be in the
+  ///                  next array element, except for short options whose
```
Comment (tautschnig) [inline src/util/cmdline.h]: `next array element in argv`

File: src/util/cmdline.h [hunk 9]
```diff
@@ -20,6 +20,58 @@ Author: Daniel Kroening, kroening@kroening.com
...
+  ///             This parameter is parsed based on optstring.
+  /// \param optstring A specification of allowed command line options.
+  ///                  This is a C string container any number of single
+  ///                  characters other than '(', ')' or ':' signifying a
+  ///                  "short" option consisting of just that character, or
+  ///                  names consisting of any characters other than ')'
+  ///                  surrounded by a matching pair of '(' and ')' signifying a
+  ///                  "long" option with the name being the string between '('
+  ///                  and ')', both of which can be optionally followed by a
+  ///                  single ':' indicating that the option takes a parameter,
```
Comment (tautschnig) [inline src/util/cmdline.h]: In my opinion, all further uses of "parameter" should be replaced by "argument" here.

**Review by tautschnig — APPROVED** [review]
Thank you for cleaning this up and adding extensive documentation. Just a bunch of minor notes.

**Pull request merged**